### PR TITLE
Adding secrets encryption with KMS key (PLAT-2263)

### PR DESCRIPTION
**Pull request by ddl-denis-parnovskiy**
The change is straightforward. Add config field, use in in eks.Cluster.

### CODE REVIEW

**Comment by ddl-denis-parnovskiy** [comment]
Discussion: should we make it optional?

Background: rule 5.3.1 requires "Ensure Kubernetes Secrets are encrypted using Customer Master Keys (CMKs) managed in AWS KMS". 

On the other hand, using default AWS keys is considered "Meduim security risk"

On the third hand I cannot find any "default AWS keys" to use for envelope encryption. 

By default the keys are encrypted at rest.

File: cdk/domino_cdk/config_template.yaml [hunk 1]
```diff
@@ -40,6 +40,7 @@ route53:
 eks:
   version: "1.19"
   private_api: false
+  secrets_encryption_key_arn: arn:aws:kms:us-west-2:890728157128:key/b9c778f3-46cc-4416-8867-4ab47d912bf6
```
Comment (ddl-denis-parnovskiy) [inline cdk/domino_cdk/config_template.yaml:43]: This is one of our keys in AWS KMS. If we go forward with making this field mandatory, we need to put some key ARN by default. There is no difference which key to use.

**Comment by Secretions** [comment]
Regarding the optionality: I think we should either make it optional, or make it so there is a version of this where we provision a KMS for the cluster for when we're not providing a pre-existing one.

Also, we don't have CI setup in this repo yet, but we do have unit tests for the configs. Would you mind validating that this (or the final version of this) passes them?

**Comment by ddl-denis-parnovskiy** [comment]
OK. I'll make it optional but add a warning message.

File: cdk/domino_cdk/config/__init__.py [hunk 2]
```diff
@@ -69,6 +70,7 @@ def config_template(
             "eks": {
                 "version": "1.19",
                 "private_api": private_api,
+                "secret_encrption_key_name": secrets_encryption_key_arn,
```
Comment (steved) [inline cdk/domino_cdk/config/__init__.py:73]:
```suggestion
                "secrets_encryption_key_arn": secrets_encryption_key_arn,
```

File: cdk/domino_cdk/config/eks.py [hunk 3]
```diff
@@ -107,6 +111,7 @@ def load(cls, ng):
     global_node_tags: Dict[str, str]
     managed_nodegroups: Dict[str, ManagedNodegroup]
     unmanaged_nodegroups: Dict[str, UnmanagedNodegroup]
+    secrets_encryption_key_arn: str
```
Comment (steved) [inline cdk/domino_cdk/config/eks.py:114]: optional?

File: cdk/domino_cdk/provisioners/eks/eks_cluster.py [hunk 4]
```diff
@@ -37,6 +39,16 @@ def provision(
...
+            secrets_encryption_key=Key.from_key_arn(
+                self.scope, "secrets_encryption_key_arn", secrets_encryption_key_arn
+            )
+            if secrets_encryption_key_arn
+            else Key(
+                self.scope,
+                f"{name}-kubernetes-secrets-envelope-key",
+                alias=f"{name}-kubernetes-secrets-envelope-key",
+                removal_policy=cdk.RemovalPolicy.DESTROY,
+            ),
```
Comment (steved) [inline cdk/domino_cdk/provisioners/eks/eks_cluster.py:51]: Can we also add `enableKeyRotation=true`?

File: cdk/domino_cdk/config/__init__.py (hunk 2 above)
Comment (ddl-denis-parnovskiy) [reply cdk/domino_cdk/config/__init__.py:73]: Let's not be confusing. It is ARN, not a name.

File: cdk/domino_cdk/config/eks.py (hunk 3 above)
Comment (ddl-denis-parnovskiy) [reply cdk/domino_cdk/config/eks.py:114]: Ah, type.

File: cdk/domino_cdk/config/__init__.py (hunk 2 above)
Comment (steved) [reply cdk/domino_cdk/config/__init__.py:73]: Sorry! I just meant to highlight the type in `secret_encryption_key_name`, totally miffed the copy-paste.

Comment (ddl-denis-parnovskiy) [reply cdk/domino_cdk/config/__init__.py:73]: Thanks!

File: cdk/domino_cdk/provisioners/eks/eks_cluster.py (hunk 4 above)
Comment (ddl-denis-parnovskiy) [reply cdk/domino_cdk/provisioners/eks/eks_cluster.py:51]: Need to try. The doc doesn't say we cannot.

Comment (ddl-denis-parnovskiy) [reply cdk/domino_cdk/provisioners/eks/eks_cluster.py:52]: This works at least during deployment.

Comment (steved) [reply cdk/domino_cdk/provisioners/eks/eks_cluster.py:52]: Thanks! It comes in the CIS compliance scans.

**Review by steved — APPROVED** [review]
LGTM, though maybe @Secretions should have final say.

File: cdk/domino_cdk/config/eks.py [hunk 5]
```diff
@@ -1,8 +1,11 @@
 from dataclasses import dataclass
-from typing import Dict, List
+from typing import Dict, List, Optional
+from logging import Logger
 
 from domino_cdk.config.util import MachineImage, check_leavins, from_loader
 
+log = Logger("domino_cdk.config")
```
Comment (Secretions) [inline cdk/domino_cdk/config/eks.py:7]: We don't actually use this and its corresponding import (was for an earlier version of the PR).

File: cdk/domino_cdk/provisioners/eks/eks_cluster.py (hunk 4 above)
Comment (Secretions) [reply cdk/domino_cdk/provisioners/eks/eks_cluster.py:52]: Just a personal preference thing, but this ternary block is big enough that I'd be inclined to evaluate it before the call to `eks.Cluster()`

File: cdk/domino_cdk/config/eks.py [hunk 6]
```diff
@@ -13,6 +16,7 @@ class EKS:
     max_nodegroup_azs: 3 - Will provision nodegroups in up to this many availability zones.
     global_node_labels: some-label: "true"  - Labels to apply to all kubernetes nodes
     global_node_tags: some-tags: "true"  - Labels to apply to all kubernetes nodes
+    secrets_encryption_key_arn: None  - KMS key name/arn to encrypt kubernetes secrets
```
Comment (Secretions) [inline cdk/domino_cdk/config/eks.py:19]: Should mention that if this is `None` we'll create the key, as this comment will get added to the generated config file (which will be replacing the static `config_template.yaml`).

**Review by Secretions — APPROVED** [review]

**Pull request merged**